common/djangoapps/mailchimp_pipeline/tests/test_tasks.py

```diff
@@ -0,0 +1,283 @@
+import json
+import requests
+import factory
+from mock import ANY, patch
+from datetime import datetime
+from django.conf import settings
+from django.test import override_settings
+from django.contrib.auth.models import User
+from django.db.models.signals import post_save
+from common.djangoapps.mailchimp_pipeline.tests.helpers import create_organization, generate_mailchimp_url
+from lms.djangoapps.onboarding.tests.factories import UserFactory, UserExtendedProfileFactory
+from openedx.core.djangoapps.content.course_overviews.models import CourseOverview
+from lms.djangoapps.onboarding.models import OrganizationMetricUpdatePrompt
+from xmodule.modulestore.tests.django_utils import ModuleStoreTestCase
+from lms.djangoapps.onboarding.models import FocusArea, OrgSector
+from lms.djangoapps.certificates import api as certificate_api
+from xmodule.modulestore.tests.factories import CourseFactory
+from student.tests.factories import UserProfileFactory
+from student.models import CourseEnrollment
+
+from mailchimp_pipeline.tasks import update_org_details_at_mailchimp, update_enrollments_completions_at_mailchimp
+from mailchimp_pipeline.helpers import get_user_active_enrollements, get_enrollements_course_short_ids
+from mailchimp_pipeline.client import MailChimpException
+from mailchimp_pipeline.signals.handlers import (
+    send_user_info_to_mailchimp,
+    sync_metric_update_prompt_with_mail_chimp,
+    update_mailchimp,
+    send_user_enrollments_to_mailchimp,
+    task_send_account_activation_email,
+    task_send_user_info_to_mailchimp,
+    send_user_course_completions_to_mailchimp
+)
+
+
+
+@override_settings(MAILCHIMP_LEARNERS_LIST_ID="test_id")
+class MailchimpPipelineTaskTestClass(ModuleStoreTestCase):
+    """
+        Tests for tasks and generic functions
+    """
+
+    ROOT_URL = "https://us12.api.mailchimp.com/3.0/"
+
+    @patch("nodebb.signals.handlers.get_current_request")
+    @patch("common.lib.mandrill_client.client.MandrillClient.send_mail")
+    def setUp(self, mocked_nodebb_request, mocked_email_sender):
+        super(MailchimpPipelineTaskTestClass, self).setUp()
+        patcher = patch('mailchimp_pipeline.client.request', autospec=True)
+        self.mock_request = patcher.start()
+        self.mock_request.return_value.status_code = 204
+        self.addCleanup(patcher.stop)
+
+        if User.objects.all().count() == 0:
+            self.user = UserFactory(is_staff=False, password='test')
+        else:
+            self.user.extended_profile = UserExtendedProfileFactory(user=self.user)
```

Comment:
why is this else necessary ?

Reply:
ModuleStoreTestCase (Base class) create user on the bases of its class Boolean variable (CREATE). So to handle both cases I have to write this if else case that if user is not already created by its Base class, create user with UserFactory (that create profile and extended profile also) otherwise if Base class has already created user (that is created without profiles) than I have to create those profiles in else case.